sqlalchemy_utils/functions/database.py

```diff
@@ -458,7 +458,11 @@ def sqlite_file_exists(database):
         return header[:16] == b'SQLite format 3\x00'
 
     url = copy(make_url(url))
-    database, url.database = url.database, None
+    try:
+        database, url.database = url.database, None
+    except AttributeError:  #SQLalchemy 1.4: url is immutable
+        database = url.database
+        url.set(database=None)
```

Comment:
Unfortunately, this won't work because `url.set()` [returns a *new* URL object](https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.engine.URL.set): it does not modify `url`.
However, `url = url.set(database=None)` won't work either because `None` cannot be assigned in the [method's current implementation](https://github.com/sqlalchemy/sqlalchemy/blob/master/lib/sqlalchemy/engine/url.py#L262).

Reply:
Hi,

For my test I added a wrapper to limit the impact in the code

Example in database_exists
```
    class UrlWrapper:
        url = {}

        def __init__(self, url):
            self.__class__.url[id(self)] = url

        def __getattr__(self, key, **kw):
            return getattr(self.url[id(self)], key, **kw)

        def __setattr__(self, key, value):
            try:
                setattr(self.url[id(self)], key, value)
            except AttributeError:
                if value is None:
                    value = ''

                self.__class__.url[id(self)] = self.url[id(self)].set(
                    **{key: value})

        def __call__(self):
            return self.url[id(self)]

    url = UrlWrapper(copy(make_url(url)))

    database, url.database = url.database, None
    engine = sa.create_engine(url())

```

Reply:
Yes, this will modify the object's attribute. But you may not really *need* to modify it. There's an explanation of the decision to make the URL object immutable in [SQLAlchemy's documentation](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#the-url-object-is-now-immutable). Setting it to `''` works, but the semantics feel off: we don't want to set the database to an empty string, we want to *unset* it.

Reply:
@ic4f

I think I have found a solution to that, please see here:

https://github.com/lucventurini/sqlalchemy-utils/commit/1f095582d0449700367a4d20e9e265074d023116

In short, I wrote a wrapper function that creates a new "URL" object (in SA>=1.4) with the database appropriately set.

Kind regards,